bundles/org.openhab.binding.energenie/src/main/java/org/openhab/binding/energenie/internal/handler/EnergeniePWMHandler.java

```diff
@@ -0,0 +1,213 @@
+
+        private final String stateResponseSearch;
+        private final int start;
+        private final int stop;
+        private final int divisor;
+        private final Unit<?> unit;
+
+        private PWMState(final String stateResponseSearch, final int start, final int stop, final int divisor,
+                final Unit<?> unit) {
+            this.stateResponseSearch = stateResponseSearch;
+            this.start = start;
+            this.stop = stop;
+            this.divisor = divisor;
+            this.unit = unit;
+        }
+
+        public String getStateResponseSearch() {
+            return stateResponseSearch;
+        }
+
+        public int getStart() {
+            return start;
+        }
+
+        public int getStop() {
+            return stop;
+        }
+
+        public int getDivisor() {
+            return divisor;
+        }
+
+        public State readState(final String loginResponseString) {
+            final int findState = loginResponseString.lastIndexOf(stateResponseSearch);
+
+            if (findState > 0) {
+                logger.trace("searchstring {} found at position {}", stateResponseSearch, findState);
+                final String slicedResponseTmp = loginResponseString.substring(findState + start, findState + stop);
+                logger.trace("transformed state response = {}", slicedResponseTmp);
+                final String[] slicedResponse = slicedResponseTmp.split(";");
+                logger.trace("transformed state response = {} - {}", slicedResponse[0], slicedResponse[1]);
+                final double value;
+
+                try {
+                    if (Double.parseDouble(slicedResponse[0]) / 1 == Double.parseDouble(slicedResponse[0])) {
+                        value = Double.parseDouble(slicedResponse[0]) / divisor;
+                    } else {
+                        value = -1.0;
+                    }
+                    return QuantityType.valueOf(value, unit);
+
+                } catch (NumberFormatException e) {
+                    logger.debug("Could not Parse State", e);
+                    return UnDefType.UNDEF;
+                }
+            } else {
+                logger.trace("searchstring '{} not found", stateResponseSearch);
+                return UnDefType.UNDEF;
+            }
+        }
+    }
+
+    private final Logger logger = LoggerFactory.getLogger(EnergeniePWMHandler.class);
+
+    private static final int HTTP_TIMEOUT_MILLISECONDS = 6000;
+
+    private @Nullable EnergenieConfiguration config;
+
+    private String host = "";
+    private String password = "";
+    private int refreshInterval;
+
+    private @Nullable ScheduledFuture<?> refreshJob;
+
+    public EnergeniePWMHandler(Thing thing) {
+        super(thing);
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (command instanceof RefreshType) {
+            try {
+                getState();
```

Comment:
This should be done asynchronously since `handleCommand` is supposed to return quickly (since it is a UI thread) and you are executing http requests with it. 
```suggestion
                scheduler.execute(this::getState);
```

Reply:
Done